In [1]:
import numpy as np
import pandas as pd
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import GridSearchCV

In [2]:
# importing the models
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.ensemble import RandomForestClassifier

In [3]:
# loading the csv data to a Pandas DataFrame
heart_data = pd.read_csv('Dataset/heart.csv')

In [4]:
# print first 5 rows of the dataset
heart_data.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  target  
0   0     1       1  
1   0     2       1  
2   0     2       1  
3   0     2       1  
4   0     2       1

In [5]:
# number of rows and columns in the dataset
heart_data.shape

(303, 14)

In [6]:
# checking for missing values
heart_data.isnull().sum()

age         0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          0
thal        0
target      0
dtype: int64

In [7]:
# checking the distribution of Target Variable
heart_data['target'].value_counts()

target
1    165
0    138
Name: count, dtype: int64

In [8]:
X = heart_data.drop(columns='target', axis=1)
Y = heart_data['target']

In [9]:
X = np.asarray(X)
Y = np.asarray(Y)

**Model Selection**

Comparing the models with Cross Validation

In [ ]:
# List of models
models = [LogisticRegression(max_iter=1000), SVC(kernel='linear'), 
          KNeighborsClassifier(), RandomForestClassifier(random_state=0)]

In [13]:
def models_cross_validation():
    for model in models:
        cv_score= cross_val_score(model, X, Y, cv=5)
        mean_accuracy = sum(cv_score / len(cv_score))
        mean_accuracy = mean_accuracy*100
        mean_accuracy = round(mean_accuracy, 2)

        print("Cross Validation accuracies for the ", model, '=', cv_score )
        print('Accuracy score of the model: ', model, '=', mean_accuracy, "%")
        print('--------------------------------------------')


In [14]:
models_cross_validation()

Cross Validation accuracies for the  LogisticRegression(max_iter=1000) = [0.80327869 0.86885246 0.85245902 0.86666667 0.75      ]
Accuracy score of the model:  LogisticRegression(max_iter=1000) = 82.83 %
--------------------------------------------
Cross Validation accuracies for the  SVC(kernel='linear') = [0.81967213 0.8852459  0.80327869 0.86666667 0.76666667]
Accuracy score of the model:  SVC(kernel='linear') = 82.83 %
--------------------------------------------
Cross Validation accuracies for the  KNeighborsClassifier() = [0.60655738 0.6557377  0.57377049 0.73333333 0.65      ]
Accuracy score of the model:  KNeighborsClassifier() = 64.39 %
--------------------------------------------
Cross Validation accuracies for the  RandomForestClassifier(random_state=0) = [0.85245902 0.90163934 0.81967213 0.81666667 0.8       ]
Accuracy score of the model:  RandomForestClassifier(random_state=0) = 83.81 %
--------------------------------------------


Random Forest Classifier 83.8% with default Hyperparameters

**Comparing Models with Different Hyperparameters**

In [30]:
# List of models
models_list = [LogisticRegression(max_iter=10000), SVC(), 
          KNeighborsClassifier(), RandomForestClassifier(random_state=0)]

In [40]:
# Creating a dictionary that contains hyperparameter values 
model_hyperparameters = {
    'log_reg_hyperparameters': {
        'C': [1,5,10,20]
    },
    'svc_hyperparameters': {
        'kernel': ['linear', 'poly', 'rbf', 'sigmoid'],
        'C': [1,5,10,20]
    },
    'knn_hyperparameters': {
        'n_neighbors' : 
            [3,5,10]
    
    },
    'random_forest_hypeparameters': {
        'n_estimators': [10, 20, 50, 100]
    }
    
}

In [41]:
type(model_hyperparameters)

dict

In [42]:
print(model_hyperparameters.keys())

dict_keys(['log_reg_hyperparameters', 'svc_hyperparameters', 'knn_hyperparameters', 'random_forest_hypeparameters'])


In [43]:
model_hyperparameters['log_reg_hyperparameters']

{'C': [1, 5, 10, 20]}

In [44]:
model_keys = list(model_hyperparameters.keys())
print(model_keys)

['log_reg_hyperparameters', 'svc_hyperparameters', 'knn_hyperparameters', 'random_forest_hypeparameters']


In [45]:
model_keys[0]

'log_reg_hyperparameters'

In [46]:
model_hyperparameters[model_keys[0]]

{'C': [1, 5, 10, 20]}

Applying GridSearchCV

In [47]:
def ModelSelection(list_of_models, hyperparameters_dictionary):

  result = []

  i = 0

  for model in list_of_models:

    key = model_keys[i]

    params = hyperparameters_dictionary[key]

    i += 1

    print(model)
    print(params)
    print('---------------------------------')


    classifier = GridSearchCV(model, params, cv=5)

    # fitting the data to classifier
    classifier.fit(X,Y)

    result.append({
        'model used' : model,
        'highest score' : classifier.best_score_,
        'best hyperparameters' : classifier.best_params_
    })

  result_dataframe = pd.DataFrame(result, columns = ['model used','highest score','best hyperparameters'])

  return result_dataframe

In [48]:
ModelSelection(models_list, model_hyperparameters)

LogisticRegression(max_iter=10000)
{'C': [1, 5, 10, 20]}
---------------------------------
SVC()
{'kernel': ['linear', 'poly', 'rbf', 'sigmoid'], 'C': [1, 5, 10, 20]}
---------------------------------
KNeighborsClassifier()
{'n_neighbors': [3, 5, 10]}
---------------------------------
RandomForestClassifier(random_state=0)
{'n_estimators': [10, 20, 50, 100]}
---------------------------------


model used  highest score  \
0      LogisticRegression(max_iter=10000)       0.831585   
1                                   SVC()       0.828306   
2                  KNeighborsClassifier()       0.643880   
3  RandomForestClassifier(random_state=0)       0.838087   

           best hyperparameters  
0                      {'C': 5}  
1  {'C': 1, 'kernel': 'linear'}  
2            {'n_neighbors': 5}  
3         {'n_estimators': 100}